### PR TITLE
Feat Login and Flash Messages

### PR DESCRIPTION
**Pull request by SanGeeky**
In this PR I add many features, of which stands out the next things:

1. Install `bcrypt` gem for password hashing
2. Add migrations to create `users`
3. Create `User` model and controller
4. Add Routes for `signup`, `singin`, `signout`
5. Add buttons to `navbar`
5. Add views `new`, `show` for users
6. Add views `new` for session
7. Create users and sessions
8. Redirects after a user signing into the app
9. Helpers for `application`
8. Authorization in controllers and Views
9. Authorize comments for logged users
10. Add Relation User and Articles
11. Add Relation User and Comments
12. Validation of Delete Comments
13. Show User information and its articles
14. Show Flash Messages when user `signup` and `signin`
15. Show validations and flash messages in the comments

### CODE REVIEW

File: app/views/articles/_list.html.erb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<!-- Articles List -->
```
Comment (mauriciosabogal) [inline app/views/articles/_list.html.erb:1]: The indentation of this file is not correct.

File: app/views/articles/show.html.erb [hunk 2]
```diff
@@ -3,41 +3,44 @@
   <%= asset_path 'post-bg.jpg' %>
 <% end %>
 <% content_for :header do %>
-  <div class="post-heading">
-    <h1><%= @article.title %></h1>
-    <span class="meta">
-      Posted by <a href="#">Rails Blog</a> on August 24, 2021
-    </span>
-  </div>
+<div class="post-heading">
```
Comment (mauriciosabogal) [inline app/views/articles/show.html.erb]: Again, the indentation of this file is incorrect.

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    session[:user_id] = nil
+    flash[:notice] = 'See you soon on Rails!'
+    redirect_back fallback_location: root_path
+  end
+
+  private
+
+  def authenticate_user
+    @user = User.find_by email: params[:email]
+    flash[:error] = 'Email not found on Rails!'
```
Comment (mauriciosabogal) [inline app/controllers/sessions_controller.rb]: You should set the error message only if the validation fails.

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    return redirect_to signin_path if @user.nil?
+
+    flash[:error] = "Incorrect password for user: #{@user.username}."
```
Comment (mauriciosabogal) [inline app/controllers/sessions_controller.rb]: You should set the error message only if the validation fails.

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  def destroy
+    session[:user_id] = nil
+    flash[:notice] = 'See you soon on Rails!'
+    redirect_back fallback_location: root_path
+  end
+
+  private
+
+  def authenticate_user
+    @user = User.find_by email: params[:email]
```
Comment (mauriciosabogal) [inline app/controllers/sessions_controller.rb]: You can move all this authentication logic to the `create` action.

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  before_action :logged_in, only: :new
+  before_action :find_user, only: :show
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    flash.now[:alert] = 'There was a problem signing up.'
```
Comment (mauriciosabogal) [inline app/controllers/users_controller.rb]: You should set the error message only if the validation fails.

**Review by mauriciosabogal — CHANGES_REQUESTED** [review]

File: app/views/articles/_list.html.erb (hunk 1 above)
Comment (SanGeeky) [reply app/views/articles/_list.html.erb:1]: Done!

File: app/views/articles/show.html.erb (hunk 2 above)
Comment (SanGeeky) [reply app/views/articles/show.html.erb]: Done!

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (SanGeeky) [reply app/controllers/sessions_controller.rb]:
```ruby
before_action :find_user, only: :create
def create
    if @user&.authenticate(params[:password])
      session[:user_id] = @user.id
      flash[:notice] = 'Welcome Back to Articles on Rails'
      redirect_to params[:previous_url]
    else
      flash[:error] = 'Incorrect email or password, please try again.'
      redirect_to signin_path
    end
end
```
Linter throws two warnings, about  `Metrics/AbcSize` so I need to add the` before_action` and  use the  Safe navigation operator  `&.`

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (SanGeeky) [reply app/controllers/sessions_controller.rb]: Same answer [session errors](https://github.com/SanGeeky/rails-training/pull/7#discussion_r622553611)

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (SanGeeky) [reply app/controllers/sessions_controller.rb]: Same answer [Answer session_controller.rb](https://github.com/SanGeeky/rails-training/pull/7#discussion_r622553611)

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (SanGeeky) [reply app/controllers/users_controller.rb]: Done !

**Review by mauriciosabogal — APPROVED** [review]

**Pull request merged**